### PR TITLE
LPD-21413 / LPD-22719 | Localize responses using Accept-Language header

### PR DESCRIPTION
**Pull request by jaimelr10**
**Description:** Allow localized responses on graphql using Accept-Language header

**Jira Tickets:** [LPD-21413](https://liferay.atlassian.net/browse/LPD-21413) / [LPD-22719](https://liferay.atlassian.net/browse/LPD-22719)

### CODE REVIEW

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java [hunk 1]
```diff
@@ -5549,6 +5622,161 @@ public void testGetScopeScopeKeyObjectEntriesPage() throws Exception {
 			_siteScopedObjectEntry1.getObjectEntryId());
 	}
 
+	@Test
+	public void testLocalizedObjectFieldWithAcceptLanguage() throws Exception {
```
Comment (sergiojimcos) [inline modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java:5626]: Just to discuss it, should be right to have a new java class called for `ObjectEntryGraphQLTest` just in order to have separated test from GraphQL and REST API? cc/ @4lejandrito @ccorreagg

**Review by sergiojimcos — CHANGES_REQUESTED** [review]
Take a view at my comments, I will review the next part

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java (hunk 1 above)
Comment (4lejandrito) [reply modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java:5626]: GrpahQL tests are autogenerated in the resource test (see [an example](https://github.com/liferay/liferay-portal/blob/63ab7283c1cdad7bff6f77f8e766ad33c8f6e566/modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-order-test/src/testIntegration/java/com/liferay/headless/commerce/admin/order/resource/v1_0/test/BaseChannelResourceTestCase.java#L556-L602)). Just rename the test to match that pattern.

Ideally the test should use the same GraphQLField classes and maybe we could place them in the HTTPTestUtil class... but we can do this at another time.

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/accept/language/AcceptLanguageImpl.java [hunk 2]
```diff
@@ -62,6 +63,10 @@ else if (acceptLanguage.equals("zh-Hant-TW")) {
 		}
 
 		try {
+			if (acceptLanguage.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (4lejandrito) [reply modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/accept/language/AcceptLanguageImpl.java:66]: Discuss this with Beni.

**Review by 4lejandrito — CHANGES_REQUESTED** [review]
See my changes and comments.

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/accept/language/AcceptLanguageImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jaimelr10) [reply modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/accept/language/AcceptLanguageImpl.java:66]: Sergio is right! I'm changing this :smile:

File: portal-test/src/com/liferay/portal/kernel/test/util/HTTPTestUtil.java [hunk 3]
```diff
@@ -100,14 +121,22 @@ public HTTPTestUtilCustomizer withCredentials(
 	}
 
 	private static Http.Options _getHttpOptions(
-		String body, String endpoint, Http.Method httpMethod) {
+		String body, String endpoint, HashMap<String, String> headers,
```
[the remark this answers is not in the document]
Comment (4lejandrito) [reply portal-test/src/com/liferay/portal/kernel/test/util/HTTPTestUtil.java:124]: You can create a task for it with `headless-tech-debt`.

**Comment by ccorreagg** [comment]
ci:test:sf

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
ci:stop

**Comment by ccorreagg** [comment]
ci:test:sf

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
ci:stop

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
ci:report:2559448194

**Comment by ccorreagg** [comment]
See the following example, given the Object Definition "Test" localizable, and a rich text object field called `richTextField`, please see the `richTextFieldRawText` value:

```shell
curl -X GET -u test@liferay.com:test http://localhost:8080/o/c/tests -H 'Accept-Language: es-ES'
```

```shell
{
  [...]
  "items" : [ {
    [...]
    "richTextField" : "<p>ES texto</p>\n",
    "richTextField_i18n" : {
      "en_US" : "<p>EN text</p>\n",
      "es_ES" : "<p>ES texto</p>\n"
    },
    "richTextFieldRawText" : "EN text"
  } ],
  "lastPage" : 1,
  "page" : 1,
  "pageSize" : 20,
  "totalCount" : 1
}
```

```shell
curl -X GET -u test@liferay.com:test http://localhost:8080/o/c/tests
```

```shell
{
  [...]
  "items" : [ {
    [...]
    "richTextField" : "<p>EN text</p>\n",
    "richTextField_i18n" : {
      "en_US" : "<p>EN text</p>\n",
      "es_ES" : "<p>ES texto</p>\n"
    },
    "richTextFieldRawText" : "EN text"
  } ],
  "lastPage" : 1,
  "page" : 1,
  "pageSize" : 20,
  "totalCount" : 1
}
```

@albertojml @4lejandrito 

This use case is covered by the `DefaultObjectEntryManagerImplTest.testUpdateObjectEntry`, is the one that is failing because of that

**Comment by ccorreagg** [comment]
Previously, the `getObjectEntry` method of the `DefaultObjectEntryManagerImpl` was returning the text fields in the language given by the `languageId` configured of the user.

However, with this approach, we are considering the preferred Locale given by the `DTOConverterContext`. This could be considered a breaking change. However, I think we should keep using the DTOConverterContext approach.

What do you think @4lejandrito ?

**Comment by ccorreagg** [comment]
> Previously, the `getObjectEntry` method of the `DefaultObjectEntryManagerImpl` was returning the text fields in the language given by the `languageId` configured of the user.
> 
> However, with this approach, we are considering the preferred Locale given by the `DTOConverterContext`. This could be considered a breaking change. However, I think we should keep using the DTOConverterContext approach.
> 
> What do you think @4lejandrito ?

I just had a talk with @4lejandrito and we agreed to mantain the DTOConverterContext approach even though it could be considered a breaking change. However, we also consider it is necessary to document the breaking change in the commit and maybe it is necessary to document it in a different way. Maybe Beni could help with that. Please @albertojml take it into account

**Comment by albertojml** [comment]
ci:test:sf

**Comment by albertojml** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
Rebased with master to fix the conflicts

**Comment by ccorreagg** [comment]
I have added some commits. Please @albertojml see https://github.com/liferay-headless/liferay-portal/pull/1961/commits/8baa7d9f8369bf65c9f52d6eeef98ef6b919fb17, https://github.com/liferay-headless/liferay-portal/pull/1961/commits/59743e433519febd243ff0c74c8a87b473bc77c8, https://github.com/liferay-headless/liferay-portal/pull/1961/commits/a097e31d37a026c270191b9c57ac9cd769a27b2f and https://github.com/liferay-headless/liferay-portal/pull/1961/commits/500971ad89571af72f84564552fd827031013fde

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
[com.liferay.object.storage.salesforce.internal.rest.manager.v1_0.test.SalesforceObjectEntryManagerImplTest](https://test-1-33.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/804229/testReport/com.liferay.object.storage.salesforce.internal.rest.manager.v1_0.test/SalesforceObjectEntryManagerImplTest) could be related @albertojml 

Could you check?

**Comment by ccorreagg** [comment]
ci:forward:force

**Comment by ccorreagg** [comment]
Tests do not seem related. Forwarding to BC

**Pull request closed without merging**